### PR TITLE
Cylindrical (GPU) LB velocity profile observable

### PR DESCRIPTION
**Pull request by KaiSzuttor**
Description of changes:
 - added new observable
 - added new observable base class for LB related implementations

PR Checklist
------------
 - [x] Tests?
   - [x] Interface
   - [ ] Core 
 - [x] Docs?

### CODE REVIEW

**Comment by fweik** [comment]
Please adapt the `AutoParameters` unit test.

File: samples/lb_profile.py [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        axis = [0, 0, 1],
+        direction = -1,
+        radius = 4.0,
+        length = 20.0)
+cylinder_boundary = espressomd.lbboundaries.LBBoundary(shape=cylinder_shape)
+system.lbboundaries.add(cylinder_boundary)
+system.integrator.run(1000)
+
+
+system.auto_update_observables.add(fluid_obs)
```
Comment (fweik) [inline samples/lb_profile.py]: Why is this needed, and what does it do?

File: src/core/observables/CylindricalLBVelocityProfile.cpp [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+*/
+#include "CylindricalLBVelocityProfile.hpp"
+#include "utils.hpp"
+#include "utils/Histogram.hpp"
+
+namespace Observables {
+
+std::vector<double> CylindricalLBVelocityProfile::
+operator()(PartCfg &partCfg) const {
+  std::vector<size_t> n_bins{{static_cast<size_t>(n_r_bins),
```
Comment (fweik) [inline src/core/observables/CylindricalLBVelocityProfile.cpp]: Please use `std::array` for fixed-size quantities.

File: src/core/observables/CylindricalLBVelocityProfile.cpp [hunk 3]
```diff
@@ -0,0 +1,73 @@
...
+                              static_cast<size_t>(n_phi_bins),
+                              static_cast<size_t>(n_z_bins)}};
+  std::vector<std::pair<double, double>> limits{
```
Comment (fweik) [inline src/core/observables/CylindricalLBVelocityProfile.cpp]: Please use `std::array` for fixed-size quantities.

File: src/core/observables/CylindricalLBVelocityProfile.cpp [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+       std::make_pair(min_z, max_z)}};
+  Utils::CylindricalHistogram<double> histogram(n_bins, 3, limits);
+#ifdef LB_GPU
+  // First collect all positions (since we want to call the LB function to
+  // get the fluid velocities only once).
+  std::vector<double> velocities(m_sample_positions.size());
+  lb_lbfluid_get_interpolated_velocity_at_positions(
+      m_sample_positions.data(), velocities.data(),
+      m_sample_positions.size() / 3);
+  ::Vector<3, double> pos_shifted, pos_cyl, velocity;
```
Comment (fweik) [inline src/core/observables/CylindricalLBVelocityProfile.cpp]: Please move those into the loop and make them `const`, like `const Vector3d shifted = {m_sample_positions ...};`

File: src/core/observables/CylindricalLBProfileObservable.hpp [hunk 5]
```diff
@@ -0,0 +1,41 @@
...
+#include "Vector.hpp"
+#include "integrate.hpp"
+#include "particle_data.hpp"
+#include "utils/coordinate_transformation.hpp"
+
+namespace Observables {
+
+class CylindricalLBProfileObservable : public CylindricalProfileObservable {
+public:
+  double sampling_delta_x=1.0, sampling_delta_y=1.0, sampling_delta_z=1.0;
```
Comment (fweik) [inline src/core/observables/CylindricalLBProfileObservable.hpp:16]: Most of this should be private.

**Review by fweik — CHANGES_REQUESTED** [review]

File: samples/lb_profile.py (hunk 1 above)
Comment (KaiSzuttor) [reply samples/lb_profile.py]: What do you mean? The whole sample? Poiseuille flow in a cylinder

File: src/core/observables/CylindricalLBProfileObservable.hpp (hunk 5 above)
Comment (KaiSzuttor) [reply src/core/observables/CylindricalLBProfileObservable.hpp:16]: You mean I should implement getter and setter?

File: src/core/observables/CylindricalLBVelocityProfile.cpp (hunk 2 above)
Comment (KaiSzuttor) [reply src/core/observables/CylindricalLBVelocityProfile.cpp]: that would need a refactoring of the histogram class

File: src/core/observables/CylindricalLBVelocityProfile.cpp (hunk 3 above)
Comment (KaiSzuttor) [reply src/core/observables/CylindricalLBVelocityProfile.cpp]: that would need a refactoring of the histogram class

File: samples/lb_profile.py (hunk 1 above)
Comment (fweik) [reply samples/lb_profile.py]: Noch, I mean `auto_update_observables`.

Comment (KaiSzuttor) [reply samples/lb_profile.py]: It auto updates the observable.

**Comment by fweik** [comment]
And what part oder your stateless observable does it update?

Am 09.03.2018 10:24 vorm. schrieb "Kai Szuttor" <notifications@github.com>:

**Review by fweik — APPROVED** [review]
